## Dog Breed Prediction

In this project, we will see how to use Keras and TensorFlow to build, train, and test a Convolutional Neural Network capable of identifying the breed of a dog in a supplied image. This is a supervised learning problem, specifically a multiclass classification problem.

We will start by connecting to Kaggle using Kaggle API which can be downloaded from your Kaggle account's settings and uploading it here(upload box).

In [2]:
# Next, install the Kaggle API client.
!pip install -q kaggle

Next we will install Kaggle API using pip installation.

Setting up Kaggle using Kaggle API.

In [4]:
# Creating directory and changing the current working directory
!mkdir dog_dataset
%cd dog_dataset

c:\Users\Ahriya Evelyn\Desktop\Data Camp Project\DogBreedPrediction\dog_dataset


To store the data we will create a new directory and make it as current working directory.

In [5]:
# Searching for dataset
!kaggle datasets list -s dogbreedidfromcomp

ref                                title                    size  lastUpdated          downloadCount  voteCount  usabilityRating  
---------------------------------  ----------------------  -----  -------------------  -------------  ---------  ---------------  
catherinehorng/dogbreedidfromcomp  dog-breed-id-from-comp  691MB  2020-06-26 03:09:05           6898         10  0.1764706        


Searching Kaggle for the required dataset using search option(-s) with title 'dogbreedidfromcomp'. We can also use different search options like searching competitions, notebooks, kernels, datasets, etc.

In [6]:
# Downloading dataset and coming out of directory
!kaggle datasets download catherinehorng/dogbreedidfromcomp
%cd ..

Dataset URL: https://www.kaggle.com/datasets/catherinehorng/dogbreedidfromcomp
License(s): unknown

c:\Users\Ahriya Evelyn\Desktop\Data Camp Project\DogBreedPrediction



  0%|          | 0.00/691M [00:00<?, ?B/s]
  0%|          | 1.00M/691M [00:01<17:05, 706kB/s]
  0%|          | 2.00M/691M [00:01<08:34, 1.40MB/s]
  0%|          | 3.00M/691M [00:01<05:22, 2.24MB/s]
  1%|          | 5.00M/691M [00:02<02:51, 4.19MB/s]
  1%|          | 7.00M/691M [00:02<01:59, 6.00MB/s]
  1%|▏         | 9.00M/691M [00:02<01:36, 7.41MB/s]
  2%|▏         | 11.0M/691M [00:02<01:22, 8.64MB/s]
  2%|▏         | 13.0M/691M [00:02<01:16, 9.24MB/s]
  2%|▏         | 15.0M/691M [00:02<01:13, 9.58MB/s]
  2%|▏         | 17.0M/691M [00:03<01:10, 10.0MB/s]
  3%|▎         | 19.0M/691M [00:03<01:09, 10.2MB/s]
  3%|▎         | 21.0M/691M [00:03<01:08, 10.2MB/s]
  3%|▎         | 23.0M/691M [00:03<01:08, 10.2MB/s]
  4%|▎         | 25.0M/691M [00:03<01:06, 10.5MB/s]
  4%|▍         | 27.0M/691M [00:04<01:05, 10.7MB/s]
  4%|▍         | 29.0M/691M [00:04<01:05, 10.7MB/s]
  4%|▍         | 31.0M/691M [00:04<01:04, 10.7MB/s]
  5%|▍         | 33.0M/691M [00:04<01:06, 10.4MB/s]
  5%|▌         | 35.0

After searching the data next step would be downloading the data into collab notebook using references found in search option.

In [8]:
# Unzipping downloaded file and removing unusable file

Importing required libraries.


In [2]:
# Important library imports
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt

from tqdm import tqdm
from keras.preprocessing import image
from sklearn.preprocessing import label_binarize
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPool2D
from keras.optimizers import Adam

We will unzip the data which is downloaded and remove the irrelevant files. 

In [3]:
# Read the labels.csv file and checking shape and records
labels_all = pd.read_csv("dog_dataset\labels.csv")
print(labels_all.shape)
labels_all.head()


(10222, 2)


<>:2: SyntaxWarning: invalid escape sequence '\l'
<>:2: SyntaxWarning: invalid escape sequence '\l'
C:\Users\Ahriya Evelyn\AppData\Local\Temp\ipykernel_15576\415472539.py:2: SyntaxWarning: invalid escape sequence '\l'
  labels_all = pd.read_csv("dog_dataset\labels.csv")


id             breed
0  000bec180eb18c7604dcecc8fe0dba07       boston_bull
1  001513dfcb2ffafc82cccf4d8bbaba97             dingo
2  001cdf01b096e06d78e9e5112d419397          pekinese
3  00214f311d5d2247d5dfe4fe24b2303d          bluetick
4  0021f9ceb3235effd7fcde7f7538ed62  golden_retriever

Loading the labels data into dataframe and viewing it. Here we analysed that labels contains 10222 rows and 2 columns.  

In [4]:
# Visualize the number of each breeds
breeds_all = labels_all["breed"]
breed_counts = breeds_all.value_counts()
breed_counts.head()

breed
scottish_deerhound      126
maltese_dog             117
afghan_hound            116
entlebucher             115
bernese_mountain_dog    114
Name: count, dtype: int64

Here we are finding out the count per class i.e. total data in each class using value_counts() function.

In [5]:
# Selecting first 3 breeds (Limitation due to computation power)
CLASS_NAMES = ['scottish_deerhound','maltese_dog','bernese_mountain_dog']
labels = labels_all[(labels_all['breed'].isin(CLASS_NAMES))]
labels = labels.reset_index()
labels.head()

index                                id                 breed
0      9  0042188c895a2f14ef64a918ed9c7b64    scottish_deerhound
1     12  00693b8bc2470375cc744a6391d397ec           maltese_dog
2     79  01e787576c003930f96c966f9c3e1d44    scottish_deerhound
3     90  022b34fd8734b39995a9f38a4f3e7b6b           maltese_dog
4    118  02d54f0dfb40038765e838459ae8c956  bernese_mountain_dog

We will work on only 3 breeds due to limited computational power. You can consider more classes as per your system computational power.

In [6]:
# Creating numpy matrix with zeros
X_data = np.zeros((len(labels), 224, 224, 3), dtype='float32')
X_data.shape

(357, 224, 224, 3)

In [7]:
# One hot encoding
Y_data = label_binarize(labels['breed'], classes = CLASS_NAMES)
Y_data

array([[1, 0, 0],
       [0, 1, 0],
       [1, 0, 0],
       ...,
       [0, 0, 1],
       [0, 1, 0],
       [0, 1, 0]])

In [8]:
# Reading and converting image to numpy array and normalizing dataset
for i in tqdm(range(len(labels))):
    img = image.load_img('dog_dataset/train/%s.jpg' % labels['id'][i], target_size=(224, 224))
    img = image.img_to_array(img)
    x = np.expand_dims(img.copy(), axis=0)
    X_data[i] = x / 255.0
    
# Printing train image and one hot encode shape & size
print('\nTrain Images shape: ',X_data.shape,' size: {:,}'.format(X_data.size))
print('One-hot encoded output shape: ',Y_data.shape,' size: {:,}'.format(Y_data.size))

100%|██████████| 357/357 [00:01<00:00, 250.76it/s]



Train Images shape:  (357, 224, 224, 3)  size: 53,738,496
One-hot encoded output shape:  (357, 3)  size: 1,071


In [9]:
X_data[0]

array([[[0.01960784, 0.02352941, 0.        ],
        [0.02745098, 0.03137255, 0.        ],
        [0.03921569, 0.04313726, 0.01176471],
        ...,
        [0.3372549 , 0.39607844, 0.0627451 ],
        [0.35686275, 0.39607844, 0.16078432],
        [0.34509805, 0.38431373, 0.14901961]],

       [[0.03137255, 0.03529412, 0.00392157],
        [0.03137255, 0.03529412, 0.00392157],
        [0.03137255, 0.03529412, 0.00392157],
        ...,
        [0.35686275, 0.41568628, 0.09803922],
        [0.35686275, 0.39607844, 0.15686275],
        [0.34901962, 0.3882353 , 0.14901961]],

       [[0.03137255, 0.03529412, 0.00392157],
        [0.02745098, 0.03137255, 0.        ],
        [0.02352941, 0.02745098, 0.        ],
        ...,
        [0.3254902 , 0.38039216, 0.07843138],
        [0.3529412 , 0.3882353 , 0.13333334],
        [0.34901962, 0.38431373, 0.12941177]],

       ...,

       [[0.20392157, 0.25882354, 0.11764706],
        [0.1254902 , 0.18039216, 0.04313726],
        [0.18039216, 0

In [10]:
import cv2

cv2.imshow("Image",X_data[0])
# Wait for 'q' key to close the window

while True:
    key = cv2.waitKey(1) & 0xFF
    if key == ord('q'):  # Press 'q' to exit
        break

-1

: 

As we are working with the classification dataset first we need to one hot encode the target value i.e. the classes. After that we will read images and convert them into numpy array and finally normalizing the array.

In [ ]:
# Building the Model
model = Sequential()

model.add(Conv2D(filters = 64, kernel_size = (5,5), activation ='relu', input_shape = (224,224,3)))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(filters = 32, kernel_size = (3,3), activation ='relu', kernel_regularizer = 'l2'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(filters = 16, kernel_size = (7,7), activation ='relu', kernel_regularizer = 'l2'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(filters = 8, kernel_size = (5,5), activation ='relu', kernel_regularizer = 'l2'))
model.add(MaxPool2D(pool_size=(2,2)))

model.add(Flatten())
model.add(Dense(128, activation = "relu", kernel_regularizer = 'l2'))
model.add(Dense(64, activation = "relu", kernel_regularizer = 'l2'))
model.add(Dense(len(CLASS_NAMES), activation = "softmax"))

model.compile(loss = 'categorical_crossentropy', optimizer = Adam(0.0001),metrics=['accuracy'])

model.summary()

Next we will create a network architecture for the model. We have used different types of layers according to their features namely Conv_2d (It is used to create a convolutional kernel that is convolved with the input layer to produce the output tensor), max_pooling2d (It is a downsampling technique which takes out the maximum value over the window defined by poolsize), flatten (It flattens the input and creates a 1D output), Dense (Dense layer produce the output as the dot product of input and kernel).


After defining the network architecture we found out the total parameters as 162,619.

In [ ]:
# Splitting the data set into training and testing data sets
X_train_and_val, X_test, Y_train_and_val, Y_test = train_test_split(X_data, Y_data, test_size = 0.1)
# Splitting the training data set into training and validation data sets
X_train, X_val, Y_train, Y_val = train_test_split(X_train_and_val, Y_train_and_val, test_size = 0.2)

After defining the network architecture we will start with splitting the test and train data then dividing train data in train and validation data. 

In [ ]:
# Training the model
epochs = 100
batch_size = 128

history = model.fit(X_train, Y_train, batch_size = batch_size, epochs = epochs, 
                    validation_data = (X_val, Y_val))

Now we will train our model on 100 epochs and a batch size of 128. You can try using more number of epochs to increase accuracy. During each epochs we can see how the model is performing by viewing the training and validation accuracy.

In [ ]:
# Plot the training history
plt.figure(figsize=(12, 5))
plt.plot(history.history['accuracy'], color='r')
plt.plot(history.history['val_accuracy'], color='b')
plt.title('Model Accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epochs')
plt.legend(['train', 'val'])

plt.show()

Here we analyse how the model is learning with each epoch in terms of accuracy.

In [ ]:
Y_pred = model.predict(X_test)
score = model.evaluate(X_test, Y_test)
print('Accuracy over the test set: \n ', round((score[1]*100), 2), '%')

We will use predict function to make predictions using this model also we are finding out the accuracy on the test set.

In [ ]:
# Plotting image to compare
plt.imshow(X_test[1,:,:,:])
plt.show()

# Finding max value from predition list and comaparing original value vs predicted
print("Originally : ",labels['breed'][np.argmax(Y_test[1])])
print("Predicted : ",labels['breed'][np.argmax(Y_pred[1])])

Here you can see image with its original and predicted label.

## Conclusion:

We started with downloading the dataset creating the model and finding out the predictions using the model. We can optimize different hyper parameters in order to tune this model for a higher accuracy. This model can be used to predict different breeds of dogs which can be further used by different NGO's working on saving animals and for educational purposes also.